codegen/src/main/resources/software/amazon/awssdk/codegen/rules2/RuleUrl.java.resource

```diff
@@ -85,4 +70,58 @@ public final class RuleUrl {
 
         return new RuleUrl(parsed.getProtocol(), parsed.getAuthority(), path, normalizedPath, isIpAddr);
     }
+
+    private static boolean isIpAddr(String host) {
+        int from = 0;
+        int segments = 0;
+        boolean done = false;
+        while (!done) {
+            int index = host.indexOf('.', from);
+            if (index == -1) {
+                if (segments != 3) {
+                    // E.g., 123.com
+                    return false;
+                }
+                index = host.length();
+                done = true;
+            } else if (segments == 3) {
+                // E.g., 1.2.3.4.5
+                return false;
+            }
+            int length = index - from;
+            if (length == 1) {
+                char ch0 = host.charAt(from);
+                if (ch0 < '0' || ch0 > '9') {
+                    return false;
+                }
+            } else if (length == 2) {
+                char ch0 = host.charAt(from);
+                char ch1 = host.charAt(from + 1);
+                if ((ch0 < '0' || ch0 > '9') || (ch1 < '0' || ch1 > '9')) {
```

Comment:
We should probably add a comment that we don't validate leading zeroes. E.g. 000.000.000.000 is not a valid IP address, but this would think it is an IP address.

That's probably fine?

Reply:
Ended up just validating that the first char is > '0'. That should do. It won't be backwards compatible but more correct, so let's do it.